### Import Libraries 

In [1]:
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Load data 

In [2]:
#load data
dataset=pd.read_csv('D:/GIS/ISDA/Data/training_data.csv')
dataset.head(4)

CHELSA_bio10_01_b1  FAPAR_mean_January  FAPAR_mean_June  precipitation  \
0                 193          621.600000       628.000000    15008.91667   
1                 173          385.458333       556.000000    15157.66667   
2                 202          401.272727       597.166667    15188.50000   
3                 295          168.326531       212.294118    15771.58333   

   elevation  landcover_cropland_precentage  landcover_class  landsat8_nir  \
0       1681                             11                4            65   
1       2201                             52                4            67   
2       1541                             22                4            64   
3        387                              0                3            49   

   landsat8_red  landsat8_swir1  landsat8_swir2   pH  sentinel2_band_2  \
0            20              62              39  6.4                28   
1            45             105              81  6.8                31   
2            31              82              52  NaN                17   
3            59              72              68  8.1                23   

   sentinel2_band_4  sentinel2_band_9  
0                37               123  
1                45                96  
2                27               146  
3                41       10000000000

In [3]:
len(dataset)

8933

### Data cleaning 

In [4]:
dataset = dataset.dropna()


In [5]:
len(dataset)

8932

In [6]:
#check column names of dataset
dataset.columns

Index(['CHELSA_bio10_01_b1', 'FAPAR_mean_January', 'FAPAR_mean_June',
       'precipitation', 'elevation', 'landcover_cropland_precentage',
       'landcover_class', 'landsat8_nir', 'landsat8_red', 'landsat8_swir1',
       'landsat8_swir2', 'pH', 'sentinel2_band_2', 'sentinel2_band_4',
       'sentinel2_band_9'],
      dtype='object')

dataset.index[49]

In [29]:
update_df = dataset.drop([dataset.index[49]])
 
update_df

update_df.to_csv(r'D:\GIS\ISDA\Data\cleandata1.csv')

### Split to X and y 

In [30]:
#split data to X and y
y = update_df["pH"]
X = update_df[["precipitation","elevation","landcover_cropland_precentage","landcover_class",
             "landsat8_nir","landsat8_red","landsat8_swir1","landsat8_swir2",
             "sentinel2_band_2","sentinel2_band_4","sentinel2_band_9"]]

### Split to train and test 

In [31]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 100)

### Regression 

In [32]:
from sklearn.linear_model import LinearRegression


In [33]:
#Instatiate model

LR = LinearRegression()

# fit model

LR.fit(X_train,y_train)

LinearRegression()

In [34]:
#predict
y_prediction =  LR.predict(X_test)
y_prediction

array([6.47792215, 6.21605303, 6.69975797, ..., 6.17871525, 6.15482392,
       5.57453955])

In [35]:
print("Intercept: ", LR.intercept_)
print("Coefficients:")
list(zip(X, LR.coef_))

Intercept:  -32.625859480230496
Coefficients:


[('precipitation', 0.0025982377915829235),
 ('elevation', 0.00030697929049930113),
 ('landcover_cropland_precentage', 0.0007281140966274118),
 ('landcover_class', -0.07188971121560907),
 ('landsat8_nir', -0.003912598336851663),
 ('landsat8_red', -0.0021185510901762508),
 ('landsat8_swir1', -0.022755870811664303),
 ('landsat8_swir2', 0.022894253292731263),
 ('sentinel2_band_2', 0.05051885468858017),
 ('sentinel2_band_4', -0.023767651066078862),
 ('sentinel2_band_9', 5.5268503779698396e-12)]

Regression equation :pH= -32.625859480230496 + (0.0026precipitation)+(0.000307elevation) + (0.000728andcover_cropland_precentage)+(-0.0719landcover_class)+(-0.003913landsat8_nir)+(-0.002119landsat8_red)+(-0.022756landsat8_swir1)+(0.022894landsat8_swir2)+(0.050519sentinel2_band_2)+(-0.023769sentinel2_band_4)+(5.526850e-12sentinel2_band_9) 


In [36]:
#Prediction of test set

y_pred_mlr= LR.predict(X_test)

In [37]:
print("Prediction for test set: {}".format(y_pred_mlr))

Prediction for test set: [6.47792215 6.21605303 6.69975797 ... 6.17871525 6.15482392 5.57453955]


In [38]:
mlr_diff = pd.DataFrame({'Actual value': y_test, 'Predicted value': y_pred_mlr})
mlr_diff.head()

Actual value  Predicted value
7189            6         6.477922
8740          6.3         6.216053
5797          3.9         6.699758
889           6.2         6.016875
8102          8.4         6.837299

### Accuracy assessment 

In [39]:
from sklearn.metrics import mean_squared_error

In [40]:
# RMSE
print(np.sqrt(mean_squared_error(y_test, y_pred_mlr)))

1.0334134821920342


 RMSE of the model is 1.033413 

In [45]:
from sklearn.metrics import r2_score

In [46]:
#R2
print(r2_score(y_test, y_pred_mlr))

0.21667401453927193


R2 is 21.67% indicating poor perfomance  